### PR TITLE
thirdparty controller add versions check

### PR DESCRIPTION
**Pull request by adohe-zz**
@liggitt I just remove version check to top level loop, also I think this loop has problem, a single TPR resource sync failure should not break the entire sync process.

---
This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/37192)

### CODE REVIEW

File: pkg/registry/extensions/rest/thirdparty_controller.go [hunk 1]
```diff
@@ -82,6 +84,12 @@ func (t *ThirdPartyController) syncResourceList(list runtime.Object) error {
 		// Loop across all schema objects for third party resources
 		for ix := range list.Items {
 			item := &list.Items[ix]
+
+			if len(item.Versions) == 0 {
+				glog.Fatalf("ThirdPartyResource %s has no defined version, skip sync", item.Name)
```
Comment (liggitt) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: this will exit the server process... probably not what you want

Comment (liggitt) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: track errors encountered while registering and aggregate them

File: pkg/registry/extensions/rest/thirdparty_controller_test.go [hunk 2]
```diff
@@ -66,7 +66,12 @@ func TestSyncAPIs(t *testing.T) {
 	resourcesNamed := func(names ...string) []expapi.ThirdPartyResource {
 		result := []expapi.ThirdPartyResource{}
 		for _, name := range names {
-			result = append(result, expapi.ThirdPartyResource{ObjectMeta: api.ObjectMeta{Name: name}})
+			result = append(result, expapi.ThirdPartyResource{
```
Comment (AdoHe) [inline pkg/registry/extensions/rest/thirdparty_controller_test.go]: sure.

**Comment by AdoHe** [comment]
@liggitt the main sync loop of TPR controller will return if one resource fail, I am considering whether we should change this also aggregate errors in such case.

**Comment by AdoHe** [comment]
@liggitt test update, ptal.

File: pkg/registry/extensions/rest/thirdparty_controller.go [hunk 3]
```diff
@@ -79,9 +79,16 @@ func (t *ThirdPartyController) syncResourceList(list runtime.Object) error {
 	existing := sets.String{}
 	switch list := list.(type) {
 	case *extensions.ThirdPartyResourceList:
+		var errs []error
```
Comment (liggitt) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: you never return this var

File: pkg/registry/extensions/rest/thirdparty_controller.go [hunk 4]
```diff
@@ -79,9 +79,16 @@ func (t *ThirdPartyController) syncResourceList(list runtime.Object) error {
...
 		// Loop across all schema objects for third party resources
 		for ix := range list.Items {
 			item := &list.Items[ix]
+
+			if len(item.Versions) == 0 {
+				errs = append(errs, fmt.Errorf("ThirdPartyResource %s has no defined version, skip sync", item.Name))
+				continue
+			}
+
```
Comment (liggitt) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: I'd expect the other error cases to append to errs instead of returning early

Comment (AdoHe) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: yes, that's just what I expect too.

**Comment by AdoHe** [comment]
@liggitt update this, please take another look.

File: pkg/master/thirdparty/thirdparty.go [hunk 5]
```diff
@@ -247,9 +247,6 @@ func (m *ThirdPartyResourceServer) InstallThirdPartyResource(rsrc *extensions.Th
 	if err != nil {
 		return err
 	}
-	if len(rsrc.Versions) == 0 {
-		return fmt.Errorf("ThirdPartyResource %s has no defined versions", rsrc.Name)
-	}
```
Comment (liggitt) [inline pkg/master/thirdparty/thirdparty.go]: I wouldn't remove this

Comment (AdoHe) [inline pkg/master/thirdparty/thirdparty.go]: I am sorry I don't get your point, why keep this?

Comment (AdoHe) [inline pkg/master/thirdparty/thirdparty.go]: yes, we should keep this, `InstallThirdPartyResource` could be called by other caller.

**Comment by AdoHe** [comment]
@liggitt please take another look.

File: pkg/registry/extensions/rest/thirdparty_controller.go [hunk 6]
```diff
@@ -79,21 +82,33 @@ func (t *ThirdPartyController) syncResourceList(list runtime.Object) error {
...
 			// place it in the set of resources that we expect, so that we don't delete it in the delete pass
 			existing.Insert(MakeThirdPartyPath(group))
 			// ensure a RESTful resource for this schema exists on the master
 			if err := t.SyncOneResource(item); err != nil {
-				return err
+				errs = append(errs, fmt.Errorf("Failed to sync ThirdPartyResource %s: %v", item.Name, err))
 			}
 		}
+		if len(errs) != 0 {
+			allErrs := errors.NewAggregate(errs)
```
Comment (liggitt) [inline pkg/registry/extensions/rest/thirdparty_controller.go]: I would expect these collected errors to be returned at the end of the function so callers know the sync wasn't fully successful

**Pull request closed without merging**